src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -255,29 +268,124 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                 index += 1;
             }
 
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
+            if (Avx2.IsSupported)
             {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<byte>.Count - 1));
+                if ((int)(byte*)index < length)
+                {
+                    nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector256<byte>.Count - 1));
+                    if ((byte*)nLength > (byte*)index)
+                    {
+                        Vector256<byte> comparison = Vector256.Create(value);
+                        do
+                        {
+                            Vector256<byte> search = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+                            int matches = Avx2.MoveMask(Avx2.CompareEqual(comparison, search));
+                            if (matches == 0)
+                            {
+                                index += Vector256<byte>.Count;
+                                continue;
+                            }
+                            // Find offset of first match
+                            else if (Bmi1.IsSupported)
```

Comment:
This would be good to extract to a helper function.

It would probably also be good for us to start centralizing these helpers, so we don't have 5 different software fallbacks spread throughout CoreFX/CoreCLR.

Reply:
Resolved in other PR